### DISCUSSION TITLE
[Fast Data][ER Schema]: how to relate two entities that don't have a direct relation.

### DISCUSSION THREAD

**Discussion by marco-basso** [mia-platform/community#124] [Q&A] [answered]
Lets assume that there are three entities: `entity-a`, `entity-b`, and `entity-c`.
`entity-a` is related to `entity-b` through `id_b` in a one to one relation.
`entity-b` is related to `entity-c` through `id_c` in a one to one relation.
I need to aggregate a value in `entity-c` starting from `entity-a`.

On the following configuration is shown how to link `entity-a` with `entity-b`. Is there a way to add an outgoing configuration on `entity-a` to get `entity-c` ?

```json
{
  "version": "N.N.N",
  "config": {
    "entity-a": {
      "outgoing": {
        "entity-b": {
          "conditions": {
            "entity_b_to_entity_a": {
              "oneToMany": false,
              "condition": {
                "id": "id_b"
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by marco-basso** [comment]
I forgot to point out that I am also interested in how to use this relation on the Single View Creator aggregation configuration.

**Comment by ThisIsDemetrio** [comment, marked as answer]
Hi @marco-basso!

I hope to have understood your question. Looking at the example of erSchema you're written, there're two solutions but it also depends from what the desired aggregation should look like.

Let's start with the _erSchema_, as this is the best you can get to connect `entity-a` with `entity-c`:
```
{
  "version": "N.N.N",
  "config": {
    "entity-a": {
      "outgoing": {
        "entity-b": {
          "conditions": {
            "entity_b_to_entity_a": {
              "oneToMany": false,
              "condition": {
                "id_a": "id_b"
              }
            }
          }
        }
      }
    },
    "entity-b": {
      outgoing": {
        "entity-c": {
          "conditions": {
            "entity_c_to_entity_b": {
              "oneToMany": false,
              "condition": {
                "id_b": "id_c"
              }
            }
          }
        }
      }
    }
  }
}
```
This is the only way, because you cannot create a direct relation between `entity-a` and `entity-c` with a condition that passes thru `entity-c`.

In your aggregation you can add a list of dependencies based also on these relationships, so you can have all of them in the same level of your mapping. Which is better explained with an example of an aggregation:
```
{
  "version": "1.1.0",
  "config": {
    "SV_CONFIG": {
      "dependencies": {
        "entity_a": {
          "type": "projection",
          "on": "_identifier"
        },
        "entity_b": {
          "type": "projection",
          "on": "entity_b_to_entity_a" // <- document of entity_b obtained via condition 
        },
        "entity_c": {
          "type": "projection",
          "on": "entity_c_to_entity_b" // <- document of entity_c obtained via condition
        },
      },
      "mapping": {
        "id_a": "entity_a.id_a",
        "id_b": "entity_b.id_b",
        "id_c": "entity_c.id_c",
        ...
      }
    }
  }
}
```
Just be careful about these two things:
- this is working if you want properties from _entity_a_, _entity_b_ and _entity_c_ at the same level of your Single View: if you need a nested structure you should create a nested mapping but you can still use conditions in your dependencies;
- I won't guarantee anything if one or more of your relations involved are _oneToMany_ because I've no idea how it would handle the array of data - but if you feel brave enough, please let us know the results (we might came up with a solution or an idea to improve the Fast Data product);

This case is tricky to me as well, since I've never encountered before personally, but I know it has been used before (and it is currently used) in some project. I'd say you should give it a try, it might work for your need.

Also, there's **another solution you can use**, [which is the *\_\_fromFile\_\_* feature](https://docs.mia-platform.eu/docs/fast_data/configuration/single_view_creator/low_code#mapping). Something like this:
```
{
  "version": "1.1.0",
  "config": {
    "SV_CONFIG": {
      "dependencies": {
        "entity_a": {
          "type": "projection",
          "on": "_identifier"
        },
        ...
      },
      "mapping": {
        "id_a": "entity_a.id_a",
        "data_from_entity_c": "__fromFile[getDataFromEntityCstartingFromEntityA.js]__",
        ...
      }
    }
  }
}
```
which gives you more freedom to get the result you want.

Hope it helped.
Much appreciated if you will let us know what you find out from this.
Cheers,
|D|

**Reply by marco-basso** [reply]
Thank you very much @ThisIsDemetrio, your answer is cristal clear and it is what I needed : )
